File: epi_judge_python/apply_permutation.py
from typing import List

from test_framework import generic_test


def apply_permutation(perm: List[int], arr: List[int]) -> None:
    for i in range(len(arr)):
        next_index = i
        while perm[next_index] >= 0:
            arr[i], arr[perm[next_index]] = arr[perm[next_index]], arr[i]
            temp = perm[next_index]
            perm[next_index] -= len(perm)
            next_index = temp

    perm[:] = [a + len(perm) for a in perm]


def apply_permutation_wrapper(perm, A):
    apply_permutation(perm, A)
    return A


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('apply_permutation.py',
                                       'apply_permutation.tsv',
                                       apply_permutation_wrapper))

File: epi_judge_cpp/apply_permutation.cc
#include <vector>

#include "test_framework/generic_test.h"

using std::vector;

void ApplyPermutation(vector<int> perm, vector<int> *A_ptr) {
    vector<int> &arr = *A_ptr;

    unsigned long long int n = perm.size();

    for (int i = 0; i < n; ++i) {
        if (perm[i] < 0) continue;
        int prev = arr[i];

        int next = i;
        while (perm[next] >= 0) {
            int putIndex = perm[next];
            perm[next] -= n;

            int temp = arr[putIndex];
            arr[putIndex] = prev;
            prev = temp;
            next = putIndex;
        }
    }

    for (int &num: perm) {
        num += n;
    }
}

vector<int> ApplyPermutationWrapper(const vector<int> &perm, vector<int> A) {
    ApplyPermutation(perm, &A);
    return A;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"perm", "A"};
    return GenericTestMain(args, "apply_permutation.cc", "apply_permutation.tsv",
                           &ApplyPermutationWrapper, DefaultComparator{},
                           param_names);
}
